src/patch_hostside.cpp

```diff
@@ -246,12 +267,19 @@ llvm::Instruction *PatchHostside::addSetKernelArgInst_byvaluevector(llvm::Instru
     args[0] = bitcast;
     args[1] = createInt32Constant(&context, allocSize);
 
+#if LLVM_VERSION_MAJOR > 4
     Function *setKernelArgHostsideBuffer = cast<Function>(M->getOrInsertFunction(
         "setKernelArgHostsideBuffer",
         Type::getVoidTy(context),
         PointerType::get(IntegerType::get(context, 8), 0),
-        IntegerType::get(context, 32),
-        NULL));
+        IntegerType::get(context, 32)));
+#else
```

Comment:
I wonder if we could do something like:
```
#if LLVM_VERSION_MAJOR > 4
#define getOrInsertFunction(M, a, b, c, d, e) \
M->getOrInsertFunction(a,b,c,d,e,0)
#else
#define getOrInsertFunction(M, a, b, c, d, e) \
M->getOrInsertFunction(a,b,c,d,e)
#endif
```
?

Reply:
I've been thinking about this too. I wasn't certain if this solution would actually lead to cleaner code or if all these macros cluttering the code would be detrimental for code clarity in the long run.
If you prefer the macro solution, I can of course change it :)

Reply:
A solution not involving macros, but that also reduces code duplication would be of course nicer.  I guess we might be able to make this macro into a standard function in fact?

Reply:
Yes, something like that might work (not yet tested):
```cpp
template<typename... ArgsTy>
Constant *getOrInsertFunction(Module* m, StringRef Name, Type *RetTy, ArgsTy... Args) {
#if LLVM_VERSION_MAJOR > 4
    return m->getOrInsertFunction(Name, RetTy, Args...);
#else
    return m->getOrInsertFunction(Name, RetTy, Args..., NULL);
#endif
}
```
Type-safe and no macro magic :) I'll implement/test that tomorrow.

Reply:
Wow, nice! :)